### PR TITLE
1

### PR DESCRIPTION
**Pull request by viktor-kniahnytskyi**
[DEMO LINK](https://viktor-kniahnytskyi.github.io/react_todo-app-add-and-delete/)

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,26 +1,37 @@
...
-        </a>
-      </p>
+    <div className="todoapp">
+      <h1 className="todoapp__title">todos</h1>
 
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+      <div className="todoapp__content">
+        <Header />
+        {!!todos.length && (
```
Comment (vadiimvooo) [inline src/App.tsx:27]:
```suggestion
        {todos.length && (
```

File: src/components/Foofer.tsx [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+            selected: status === Status.Completed,
+          })}
+          data-cy="FilterLinkCompleted"
+          onClick={() => setStatus(Status.Completed)}
+        >
+          Completed
+        </a>
+      </nav>
+
+      {/* this button should be disabled if there are no completed todos */}
```
Comment (vadiimvooo) [inline src/components/Foofer.tsx:65]:
```suggestion
```

**Review by vadiimvooo — CHANGES_REQUESTED** [review]
Well done. You have to deploy your solution. Also please pass all the tests. Feel free to ask for help in fe_chat if you need some.

File: src/components/Foofer.tsx [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+import { ErrText } from '../types/ErrText';
+
+export const Foofer: React.FC = () => {
+  const { todos, status, setStatus, deleteTodo, setErrMessage } = useTodos();
+  const itemsLeft = todos.filter(el => !el.completed).length;
+  const completedTodos = todos.filter(todo => todo.completed);
+
+  const clearCompleted = async () => {
+    try {
+      await Promise.all(completedTodos.map(todo => deleteTodo(todo.id)));
```
Comment (vadiimvooo) [inline src/components/Foofer.tsx:14]: This Promise all can cause test failure, please try to use just for loop and delete every todo individually.

**Review by vadiimvooo — CHANGES_REQUESTED** [review]
Good job. Try to fix the comment and pass the tests. Feel free to ask for help in chat.

File: src/App.tsx [hunk 4]
```diff
@@ -1,26 +1,209 @@
...
+  const filteredTodos = todos.filter(task => {
+    switch (status) {
+      case Status.Active:
+        return !task.completed;
+      case Status.Completed:
+        return task.completed;
+      case Status.All:
+      default:
+        return true;
+    }
```
Comment (volodymyr-soltys97) [inline src/App.tsx:60]:
```suggestion
    switch (status) {
      case Status.Active:
        return !task.completed;
        
      case Status.Completed:
        return task.completed;
        
      case Status.All:
      default:
        return true;
    }
```

File: src/App.tsx [hunk 5]
```diff
@@ -1,26 +1,209 @@
...
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
+  const itemsLeft = todos.filter(({ completed }) => !completed).length;
+  const haveCompletedTodos = todos.some(({ completed }) => completed);
+  const allCompleted = todos.every(({ completed }) => completed);
+
+  const filteredTodos = todos.filter(task => {
```
Comment (volodymyr-soltys97) [inline src/App.tsx:51]:
```suggestion
  const filteredTodos = todos.filter(todo => {
```

File: src/App.tsx [hunk 6]
```diff
@@ -1,26 +1,209 @@
...
+          errorMessage={errorMessage}
+        />
+        <TodoList
+          todos={filteredTodos}
+          onDeleteTodo={deleteTodo}
+          updateTodo={updateTodo}
+          tempTodo={tempTodo}
+          loading={loading}
+        />
+        {todos.length > 0 && (
```
Comment (volodymyr-soltys97) [inline src/App.tsx:177]:
```suggestion
        {!!todos.length && (
```

File: src/components/Foofer.tsx [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+      setTimeout(() => setErrMessage(ErrText.NoErr), 3000);
+    }
+  };
+
+  return (
+    <footer className="todoapp__footer" data-cy="Footer">
+      <span className="todo-count" data-cy="TodosCounter">
+        {itemsLeft}
+        {itemsLeft > 1 ? ' items ' : ' item '}
+        left
```
Comment (volodymyr-soltys97) [inline src/components/Foofer.tsx:27]: Move this checking to the helper variable and use it here

File: src/components/TodoItem.tsx [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+            onBlur={handleSaveEdit}
+          />
+        </form>
+      ) : (
+        <span
+          data-cy="TodoTitle"
+          className={cn('todo__title', { hidden: isEditing })}
+          onClick={handleEditToggle}
+        >
+          {todo.title}
```
Comment (volodymyr-soltys97) [inline src/components/TodoItem.tsx:87]: Use destruction for `todo`

File: src/components/TodoItem.tsx [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+        </span>
+      )}
+
+      <div
+        data-cy="TodoLoader"
+        className={`modal overlay ${!todo.id || (loading && todo.id === selectedTodo?.id) ? 'is-active' : ''}`}
```
Comment (volodymyr-soltys97) [inline src/components/TodoItem.tsx:93]: Use classnames library here for add classes with condition

File: src/components/TodoItem.tsx [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+      </div>
+
+      <button
+        type="button"
+        className="todo__remove"
+        data-cy="TodoDelete"
+        onClick={() => {
+          onDeleteTodo(todo.id);
+          setSelectedTodo(todo);
+        }}
```
Comment (volodymyr-soltys97) [inline src/components/TodoItem.tsx:106]: Move this logic from jsx to the helper function and use it here

**Review by volodymyr-soltys97 — CHANGES_REQUESTED** [review]
Good job 👍 
Let's make your code better

**Pull request closed without merging**